pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java

```diff
@@ -258,16 +258,59 @@ public void incrementPublishCount(int numOfMessages, long msgSizeInBytes) {
      @Override
     public void resetPublishCountAndEnableReadIfRequired() {
         if (this.publishRateLimiter.resetPublishCount()) {
-            enableProduerRead();
+            enableProducerRead();
         }
     }
 
      /**
      * it sets cnx auto-readable if producer's cnx is disabled due to publish-throttling
      */
-    protected void enableProduerRead() {
+    protected void enableProducerRead() {
         if (producers != null) {
-            producers.forEach(producer -> producer.getCnx().enableCnxAutoRead());
+            producers.values().forEach(producer -> producer.getCnx().enableCnxAutoRead());
+        }
+    }
+
+    protected void checkTopicFenced() throws BrokerServiceException {
+        if (isFenced) {
+            log.warn("[{}] Attempting to add producer to a fenced topic", topic);
+            throw new BrokerServiceException.TopicFencedException("Topic is temporarily unavailable");
+        }
+    }
+
+    protected void internalAddProducer(Producer producer) throws BrokerServiceException {
+        if (isProducersExceeded()) {
+            log.warn("[{}] Attempting to add producer to topic which reached max producers limit", topic);
+            throw new BrokerServiceException.ProducerBusyException("Topic reached max producers limit");
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] {} Got request to create producer ", topic, producer.getProducerName());
+        }
+
+        Producer existProducer = producers.putIfAbsent(producer.getProducerName(), producer);
+        if (existProducer != null) {
+            tryOverwriteOldProducer(existProducer, producer);
+        }
+    }
+
+    private void tryOverwriteOldProducer(Producer oldProducer, Producer newProducer)
+            throws BrokerServiceException {
+        boolean canOverwrite = false;
+        if (oldProducer.equals(newProducer) && !oldProducer.isUserProvidedProducerName()
+                && !newProducer.isUserProvidedProducerName() && newProducer.getEpoch() > oldProducer.getEpoch()) {
+            oldProducer.close();
+            canOverwrite = true;
+        }
+        if (canOverwrite) {
```

Comment:
nit: this can be simplified with

```
if (canOverwrite && !producers.replace(newProducer.getProducerName(), oldProducer, newProducer)) {
      throw new BrokerServiceException.NamingException(
                    "Producer with name '" + newProducer.getProducerName() + "' is already connected 
}
```

Reply:
May be if can simplified with 
```
if (!canOverwrite || !producers.replace(newProducer.getProducerName(), oldProducer, newProducer)) {
    throw new BrokerServiceException.NamingException(
                    "Producer with name '" + newProducer.getProducerName() + "' is already connected")
}
```